### DISCUSSION TITLE
Shared Data - music library updates not showing in NAS library on 'client' machines

### DISCUSSION THREAD

**Discussion by ceejayemm** [rern/rAudio#802] [Bug Reports]
RPi 3A+ Update 20220204

It seems that the Shared Data facilities are now working correctly - except that on the 'client' systems the music library is not being fully updated as detailed below.

I have added new albums via the 'master' system which show up on this system under the NAS share in the Library and also in the Albums section as expected.  The albums are playable on the 'master' system.

On the 'client' systems the new albums appear in the 'shared' Albums area but do not appear in the NAS folders - so they are not playable on the 'client' systems.

In order to get the NAS folder structure updated so that albums will play it is still necessary to run a Database Update on the top level NAS folder. Is this expected behaviour ?

Regards.

Chris

**Comment by ceejayemm** [comment]
@rern 

With a bit of time on my hands recently, I think I have a solution to the problems outlined above. After reading up on mpd here: 

https://mpd.readthedocs.io/en/stable/user.html#satellite-setup

Using my Office instance, an RPi 3B+ with a wired gigabit network connection, as the base for the music library I configured rAudio-1 to use the standard 'Shared Data' storage feature. I ran a Full Library Update on this instance to ensure I had the latest music library items and, once complete, checked this out to make sure it was OK and everything worked as I would expect on this instance - it did. This is now my 'Server' installation.

Moving over to the Loft instance, which is now being used as a 'Client', I also configured it to use the 'Shared Data' feature, but then modified the /etc/mpd.conf file (after taking a backup so I could revert easily if necessary) as follows:

I commented out the existing db_file setting
I added a 'satellite database' block which pointed to the Office installation (192.1.68.86.31)

as shown below:

bind_to_address        "0.0.0.0"
#db_file                "/srv/http/data/mpd/mpd.db"
music_directory        "/mnt/MPD"
playlist_directory     "/var/lib/mpd/playlists"
replaygain             "off"
state_file             "/var/lib/mpd/mpdstate"
user                   "mpd"

database {
    plugin "proxy"
    host "192.168.86.31"
}

input {
        plugin         "curl"
}

decoder {
        plugin         "ffmpeg"
        enabled        "yes"
}

resampler {
        plugin         "soxr"
        quality        "very high"
}

audio_output {
        name           "On-board Headphones"
        device         "hw:0,0"
        type           "alsa"
        auto_resample  "no"
        auto_format    "no"
        mixer_type     "none"
}

I rebooted the Loft instance and checked out the Library, it was showing a full copy of Library (and the same as the Office instance) immediately, all albums, artists, counts etc were present and ready to use.

To make this available through the rAudio-1 the existing System / Shared Data option would need to be changed to add extra options:

Designate whether this is a server or client instance. 
Additionally if a client, the Name/IP Number of the associated server

If this instance is to be a 'Server' then no changes need to be made to the existing setup scripts. However if it is to be a 'Client' then it will be necessary to use the user supplied information to modify the /etc/mpd.conf file accordingly along the lines shown above.

Just 'food for thought' but I hope it helps.

Chris

**Reply by rern** [reply]
Hi Chris,

Please try mounting music share before enable Shared Data.

rern

**Comment by ceejayemm** [comment]
@rern 

My Shared Data folder is just another folder on the SAME NFS share as my music library. So if the Music Library is mounted so is the Shared Data folder.

Regards

Chris

**Comment by rern** [comment]
MPD needs a restart to load the updated database.
Please try the UPDATE branch:
- Disable Shared Data
- Re-enable

**Reply by ceejayemm** [reply]
@rern 

Do I have to do this every time the main library is updated to get the new items to the client systems ? Or is it a one-off to implement new Shared Data functionality ?

Chris

**Reply by rern** [reply]
Have you tried it?

**Reply by ceejayemm** [reply]
Yes, I tried it briefly yesterday but I couldn't see any difference. I am away from home until later in the week and will give it another try when I return home. 

Chris

**Reply by feebb** [reply]
If the command `systemctl restart mpd` is executed on the Shared Data Client, the Client Library will update to match that of the Shared Data Server. Perhaps a button placed on either the System/Shared Data or the Library Update page that the user could press on the Client whenever the Library on the Shared Data Server was updated would help with this issue.

**Reply by ceejayemm** [reply]
@rern 

I have 

- disabled the Shared Data option on both my server and client systems. 
- updated the Library on the server system to ensure I have the latest album data in my Library. 
- deleted the Shared Data folders from the NFS share 
- re-enabled Shared Data on the server system and allowed its database to be used to recreate the Shared Data folders. 
- reconnected the client system to the Shared Data NFS share
- checked the Library on the client system, no sign of the new albums
- used the systemctl restart mpd command on the client system as suggested by @feebb above, still no sign of the new albums
- rebooted client system, still no sign of new albums
- ran database update (NAS page) on client system, new albums appeared but library/database on server system screwed up (only corrected by rebuilding Library on Server system.

Overall, I don't think this was a success.  For the time being I think I will stay with my modification to use a satellite database as detailed above as this seems much more straightforward, in keeping with my idea of a client/server shared data setup and seems to work without any problems.

Chris

**Reply by feebb** [reply]
Like @ceejayemm, I had a number of issues setting up shared data in a similar configuration (A server system in a downstairs room and a client in an upstairs loft). When my initial set up didn't work, I followed pretty much the same routine, but with the following differences, written in **bold** 

I
- disabled the Shared Data option on both my server and client systems.

- updated the Library on the server system to ensure I have the latest album data in my Library.

- deleted the Shared Data folders from the NFS share

- re-enabled Shared Data on the server system and allowed its database to be used to recreate the Shared Data folders.

- reconnected the client system to the Shared Data NFS share

- checked the Library on the client system, no sign of the new albums **At this point, I could actually see the folders from the share after rebooting the client  system, but I couldn't play them. So I added the NAS Folders where the music was physically located under Storage, but made sure that the Library would not update when the NAS Folder was added. I was now able to play music on the Client system from the Library that was identical to that on the Server system**

- **When I added new music to the NAS Shared Folder and updated the Server Music Library, the Client Library did not reflect those changes unless I rebooted.**

- **Since mpd needed to be restarted**, I used the systemctl restart mpd command on the client system** and the Client Library showed the new music.** 

- **I did not run** database update (NAS page) on client system **because I assumed that would result in creating a new database on the Client system, which would defeat the purpose of having Shared Data**

I don't know if this will work generally, since we may have slightly different versions of rAudio. I have not updated from the Update branch, for example in a while. 

One other issue with the Shared Data setup is the inability to review or change the settings once it is set up. The way it works now is you have to completely start over even if you just want to make sure you didn't enter the wrong info in the Shared Data section. 

I really hope the Shared Data feature can be straightened out, since it really is a great addition to rAudio. Many thanks to @rern for all his excellent work on this project!
